### PR TITLE
Better telemetry & handling of disconnect flows

### PR DESCRIPTION
**Pull request by vladsud**
This change covers one of the problems identified in Issue #3896 - some disconnect events do not carry proper reason, pointing out to some logical errors. While I can't pinpoint to particular flaw in current code, this change adds additional telemetry to understand better the problem, but also gets rid of reentrancy by making sure that all event handlers properly removed on various disconnect / error flows.

### CODE REVIEW

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 1]
```diff
@@ -305,16 +305,40 @@ export class DocumentDeltaConnection
...
+        // still send disconnect_document which is allowed; this is only intended to prevent normal messages from
+        // being emitted.
+        if (this.closed) {
+            return;
+        }
+        this.closed = true;
+
         this.removeTrackedListeners(false);
+        this.disconnectCore(socketProtocolError, reason);
+        this.removeAllListeners();
```
Comment (ChumpChief) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:337]: I'm not a fan of removing event listeners in this manner in general, it leaves the listeners unaware that their listener has been modified which can be hard to account for in internal state.

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:337]: Is your comment applicable only to this function, or in general? I believe close() method has to unregister, otherwise this memory is leaked - that's standard flow we use in other places on closure.

This particular case is a bit weird, because I do not understand when disconnect() can be called not through close() (using function names of this PR), when there are actually callbacks installed by DeltaManager. So to some extent, that's safety mechanism. Also unregistering listeners should follow removeTrackedListeners(false), as no events are going to be fired anyway (we unsubscribed from all events on socket).

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts [hunk 2]
```diff
@@ -340,30 +340,20 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection impleme
...
-            OdspDocumentDeltaConnection.removeSocketIoReference(key, socketProtocolError, reason);
+        OdspDocumentDeltaConnection.removeSocketIoReference(key, socketProtocolError, reason);
 
-            // RemoveSocketIoReference() above raises "disconnect" event on socket for socketProtocolError === true
-            // If it's not critical error, we want to raise event on this object only.
-            this.emit("disconnect", reason);
-        }
+        // RemoveSocketIoReference() above raises "disconnect" event on socket for socketProtocolError === true
+        // If it's not critical error, we want to raise event on this object only.
+        this.emit("disconnect", reason);
```
Comment (ChumpChief) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:357]: I think the call in `disconnect()` to `removeTrackedListeners()` will remove `"disconnect"` listeners before this has a chance to emit?

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:337]: In general.  My preference is for listeners to always unregister themselves when it's appropriate, rather than the emitter doing the unregister.

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts (hunk 2 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:357]: Nevermind, I get it now - was confusing the socket object and the document delta connection object here.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:357]: The issue has description of these 3 paths:
1. If the call comes from close(), then yes, it removes all listeners - we do not want reentrancy here - this connection is being silently closed. 
2. If the call comes from early handlers, then we do raise an event, and then remove all listeners. This path should not happen (i.e. there should be no listeners), but somehow (based on telemetry) we do have listeners. Propagating reason will help us understand from where call happens.
3. "error" handler is special, as we do call disconnect in it, but by this time client should have closed connection. If it did not, it works like # 2 above - raises event and closes connection.

**Comment by vladsud** [comment]
Copying relevant part from Issue #3896:

> One interesting observation: we should never ever get "client closing connection" reason for disconnect.
> This is coming from OdspDocumentDeltaConnection.disconnect() that is called in these 3 workflows:
> 
> 1. From DeltaManager (through DeltaConnection.close() flow), but we already emitted disconnect event with real reason and reset this.conneciton to undefined to protect against reentrancy.
> 2. From various early handlers installed via addConnectionListener() that should be removed on successful connection. If any of these handlers is called , that means we do not have connection, i.e. no DeltaManager handlers had a chance to be installed and produce this telemetry.
> 3. "error" handler - "error" event is raised and that should be the reason for "disconnect",, events raised from disconnect() called later should not reach DeltaManager.
> 
> None of these should result in this message ever getting to telemetry.

I was able once to hit # 2 above. I still do not understand how this is possible :( Inspecting code, I still do not see a way how that happens, unless there is some logic error at EventEmitter level

**Comment by vladsud** [comment]
One case how we get "client closing connection":
There are two users of socket (ODSP driver).
One fully connected, another connecting.
When "disconnect" event happens, if the event handler for second connection runs first (i.e. events are served in LIFO order), then this.disconnect(true) will be called (using old names) and "disconnect" event would be raised for first connection user with that cryptic message. This change addresses it - proper message would be delivered to all clients.

Also after recent iteration of cleanup, DeltaConnection can be removed as non-functional wrapper - opened an issue for that to follow up

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 1 above)
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts]: Take a look - I'm not sure latest iteration is better, but now the caller calls removeAllListeners()

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 3]
```diff
@@ -305,16 +315,41 @@ export class DocumentDeltaConnection
         this.submitManager.add("submitSignal", [message]);
     }
 
+    /**
+     * Disconnect from the websocket, and permanently disable this DocumentDeltaConnection.
+     */
+    public close() {
```
Comment (ChumpChief) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: In most of our objects the `close()` method permanently disables/disposes the object (the comment here implies that also).  However, it doesn't look like we enforce any permanent disable here and I see `canRetry` is `true` below.  Maybe rename to keep the semantic of "close" consistent in the repo.

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: I'm not sure what we can improve here. Could you please clarify?
As a general statement, subclassing can break any invariant for any class, I do not think there is a cure against that :)

Comment (ChumpChief) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: I think I don't really understand why `disconnectCore` was split out from `disconnect`, I only see `disconnectCore` getting overridden with this change, is there a scenario for overriding `disconnect` rather than `disconnectCore` with this change?

If not, then I think merging `disconnect` into `close` but keeping `disconnectCore` split out brings it back in line with our other `close` methods.  Also maybe `canRetry` should be false?

File: packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts [hunk 4]
```diff
@@ -204,7 +204,7 @@ export class ReplayFileDeltaConnection
         debug("dropping the outbound signal and wait for response");
     }
 
-    public disconnect() {
+    public close() {
         debug("no implementation for disconnect...");
```
Comment (ChumpChief) [inline packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts:208]:
```suggestion
        debug("no implementation for close...");
```

File: packages/drivers/replay-driver/src/replayDocumentDeltaConnection.ts [hunk 5]
```diff
@@ -294,7 +294,7 @@ export class ReplayDocumentDeltaConnection
         debug("dropping the outbound signal and wait for response");
     }
 
-    public disconnect() {
+    public close() {
         debug("no implementation for disconnect...");
```
Comment (ChumpChief) [inline packages/drivers/replay-driver/src/replayDocumentDeltaConnection.ts:298]:
```suggestion
        debug("no implementation for close...");
```

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 3 above)
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: The only difference I split disconnect & close is to be more assertive about need to provide reason for disconnect. Same can be done with single method and default args, but splitting seems cleaner - close() is for external users, disconnect() for internal and they have to provide reason, it can't be some default reason. disconnectCore is the one that is supposed to be overwritten only.
I can merge them back together if you feel like it's more readable / less likely for people to make mistakes.

File: packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts (hunk 4 above)
Comment (vladsud) [reply packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts:208]: Thanks! I'd rather remove all debug() statements altogether - I believe they are great for initial (local) debugging, but shipping code should use telemetry...

File: packages/loader/test-loader-utils/src/mockDocumentDeltaConnection.ts [hunk 6]
```diff
@@ -73,7 +73,7 @@ export class MockDocumentDeltaConnection
             this.submitSignalHandler(message);
         }
     }
-    public disconnect(reason?: string) {
+    public close(reason?: string) {
         this.emit("disconnect", reason ?? "mock disconnect called");
```
Comment (ChumpChief) [inline packages/loader/test-loader-utils/src/mockDocumentDeltaConnection.ts:77]:
```suggestion
        this.emit("disconnect", reason ?? "mock close called");
```

File: packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts (hunk 4 above)
Comment (ChumpChief) [reply packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts:208]: Either seems fine to me, just want to avoid stale terminology

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 3 above)
Comment (ChumpChief) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: Agree the split is good, just suggesting 2 methods rather than 3:

```typescript
public close() // don't override this, sets the closed flag
protected disconnect(socketProtocolError: boolean, reason: DriverError) // good to override, by default disconnects the socket
```

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: Are you sure? :)
I really do not want to duplicate that code across 2 implementations:
```
    protected disconnect(socketProtocolError: boolean, reason: DriverError) {
        // We set the closed flag as a part of the contract for overriding the disconnect method. This is used by
        // DocumentDeltaConnection to determine if emitting on the socket is allowed, which is important since
        // OdspDocumentDeltaConnection reuses the socket rather than truly disconnecting it.  Note that below we may
        // still send disconnect_document which is allowed; this is only intended to prevent normal messages from
        // being emitted.
        if (this.closed) {
            return;
        }
        this.closed = true;

        this.removeTrackedListeners(false);
```

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: Or are you suggesting moving it into close?

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: this.close handling is probably fine being moved to close, this.removeTrackedListeners(false) call would need to be duplicated that way

Comment (ChumpChief) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: The latter, move it into close (so then close is the one setting the closed flag, ensuring the object is permanently disabled even if the protected method is overridden)

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:321]: Take a look at latest iteration - I reorganized how we react to various errors by introducing fail() callback.
The weak spot - if someone calls disconnect() directly, they are not clearing callbacks...

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 7]
```diff
@@ -305,16 +315,37 @@ export class DocumentDeltaConnection
...
+            "client closing connection",
+            true, // canRetry
+        );
+        this.removeTrackedListeners(false);
+        this.disconnect(false, reason);
+    }
+
     /**
      * Disconnect from the websocket, and permanently disable this DocumentDeltaConnection.  Subclasses which
      * override this method must set the "closed" flag after disconnecting.
```
Comment (ChumpChief) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:342]: Comment needs update, now that closed flag is set in `close`

File: packages/loader/container-loader/src/deltaManager.ts [hunk 8]
```diff
@@ -812,7 +812,9 @@ export class DeltaManager
         this.stopSequenceNumberUpdate();
 
         // This raises "disconnect" event
-        this.disconnectFromDeltaStream(error !== undefined ? `${error.message}` : "Container closed");
+        if (this.connection !== undefined) {
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/deltaManager.ts:815]: Why move this check outside of `disconnectFromDeltaStream`?  Seems like this makes it easier to call it incorrectly?

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:815]: I can move it in. I change it mostly because old design assumed reentrancy (i.e. event handlers being called on connection that is either not current or in process of being closed). This PR removes any reentrancy (by unregistering handlers on close) and asserts on reentrancy. So any handler calling disconnectFromDeltaStream() on a connection that is not valid/current is violation.

That said, call calls to disconnectFromDeltaStream() from handlers comes through reconnectOnError() that is already checks for reentrancy, so it does not really matter what disconnectFromDeltaStream() does.

File: packages/loader/container-loader/src/deltaConnection.ts [hunk 9]
```diff
@@ -74,35 +60,19 @@ export class DeltaConnection
...
-        });
-
         this.on("newListener", (event: string, listener: (...args: any[]) => void) => {
             // Register for the event on connection
             // A number of events that are pass-through.
             // Note that we delay subscribing to op / signal on purpose, as
             // that is used as a signal in DocumentDeltaConnection to know if anyone has subscribed
             // to these events, and thus stop accumulating ops / signals in early handlers.
             // See DocumentDeltaConnection.initialMessages() implementation for details.
-            if (this.forwardEvents.includes(event)) {
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/deltaConnection.ts:95]: Although I also don't like the event forwarding logic here, the forwardEvents check protects against passing through events that don't match the `IDocumentDeltaConnectionEvents` set.  Perhaps we should keep it until we get rid of the event forwarding?

File: packages/loader/container-loader/src/deltaManager.ts (hunk 8 above)
Comment (ChumpChief) [reply packages/loader/container-loader/src/deltaManager.ts]: I'd prob prefer it inside, or else at least add a note in the method doc that a precondition for calling is being in the connected state.  Calling disconnect when in a disconnected state feels like it should be a safe no-op.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 10]
```diff
@@ -156,6 +164,8 @@ export class DocumentDeltaConnection
             });
 
         this.on("newListener", (event, listener) => {
+            assert(!this.closed, "can't add listeners after close");
```
Comment (ChumpChief) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:167]: Event registration is a public interface, but the closed property is protected so it's impossible for a caller to know if their event registration will throw.  It's also atypical for event registration to throw (at least I can't think of any other precedent), so I'd prefer not to assert here.

Comment (ChumpChief) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:167]: Better option would be to silently not register I think - it's a bit of a weird one because of the event forwarding thing.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 11]
```diff
@@ -323,25 +353,29 @@ export class DocumentDeltaConnection
         // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
         this.socket.on("signal", this.earlySignalHandler!);
 
+        let success = false;
+
         this._details = await new Promise<IConnected>((resolve, reject) => {
+            const fail = (socketProtocolError: boolean, err: DriverError) => {
```
Comment (ChumpChief) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:359]: Should this also set `this.closed = true`?

Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:359]: Yes, I think it makes sense, though it's a bit funny, as we are getting more and more code duplication between close() and fail(0 methods :)

File: packages/loader/container-loader/src/deltaConnection.ts (hunk 9 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaConnection.ts:95]: I'd not spend any time here, but rather remove all this code in next PR right after this one.
That said, it's interesting question, as we do not do these kids of checks anywhere else, so it begs the question - should we? including in DocumentDeltaConnection layer.
And can be we it more generically in TypedEventEmitter?

Opened https://github.com/microsoft/FluidFramework/issues/3930 to track it.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 10 above)
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:167]: I like not-registering. Changing.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 12]
```diff
@@ -1058,6 +1057,8 @@ export class DeltaManager
         this._outbound.clear();
         this.emit("disconnect", reason);
 
+        // Avoid re-entrncy - remove all listeners before closing!
+        connection.removeAllListeners();
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/deltaManager.ts:1061]: Nit: I mentioned elsewhere that I'm not a fan of `removeAllListeners()`, but it's probably benign here since I don't think we hand out the connection object.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 13]
```diff
@@ -1058,6 +1057,8 @@ export class DeltaManager
         this._outbound.clear();
         this.emit("disconnect", reason);
 
+        // Avoid re-entrncy - remove all listeners before closing!
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/deltaManager.ts:1060]:
```suggestion
        // Avoid re-entrancy - remove all listeners before closing!
```

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts [hunk 14]
```diff
@@ -340,30 +340,20 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection impleme
...
```
Comment (ChumpChief) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:357]: I think currently this causes the DeltaConnection to close itself, but the handler is removed now.  Is this behavior important to preserve?

**Review by ChumpChief — APPROVED** [review]
Nice improvements (def like the split of close/disconnect and the fail callback) but still very hard to follow the eventing/state transitions.  I'll spend some time here on the driver work.

File: packages/loader/container-loader/src/deltaConnection.ts (hunk 9 above)
Comment (ChumpChief) [reply packages/loader/container-loader/src/deltaConnection.ts:95]: It's specific to the newListener and event forwarding behavior, so most of our classes have no problem.  Most event emitters can know exactly what events they emit (because they are the ones doing the emitting).  But with the re-forwarding, it becomes possible for DeltaConnection to emit any event that the document deltaconnection emits, which it can't know and will vary based on the document deltaconnection implementation.  The forwardEvents check is the current mechanism by which the DeltaConnection can know again exactly what events it emits.

File: packages/loader/container-loader/src/deltaManager.ts (hunk 12 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:1061]: Correct, I think code would be more convoluted if we keep remembering callbacks and unregister them one by one - more chanced to screw it up - all of this logic is already too complex to easily review and spot bugs :)

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts (hunk 14 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:357]: DeltaConnection is benign pass-through (in this PR), it does not do anything.
Also handlers are removed on socket (prior to getting here), not on DocumentDeltaConnection instance (or derived class) - this is now (after your feedback) moved to be responsibility of the caller/user of these classes.

Please let me know if I got it wrong.

File: packages/loader/container-loader/src/deltaConnection.ts (hunk 9 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaConnection.ts:95]: There are two problems:
1. How do we make sure that caller can register only on events that this class supports. This is the problem of TypedEventEmitter. Code like it exists here (before my changes) should not exist, it should be solved centrally, without creating new sources of information (list of events). The problem shows up today as "error" missing on signature of IDocumentDeltaConnectionEvents
2. How do we ensure in cases like that (forwarder) that adapter layer's (DeltaConnection) event set is strict subset of class it wraps. TypedEventEmitter (used by IDocumentDeltaConnection implementations) would help enforce it at runtime, but there is no solution for compile time (likely similar to # 1 - we are talking mostly about runtime checks, due to having generic on() handler).

I do not see an easy solution for # 2, and there never was one here.
\# 1 is more generic and is strictly a problem of TypedEventEmitter - I'm against of every TypedEventEmitter consumer maintaining duplicate list of events it knows about.

Comment (vladsud) [reply packages/loader/container-loader/src/deltaConnection.ts:95]: I'll merge this PR, as this point is very moot either way - I'll raise PR right away to delete all this code :)
